File: api/api-base/src/test/java/com/xpay/common/api/test/RequestUtil.java
package com.xpay.common.api.test;

import com.xpay.api.base.params.RequestParam;
import com.xpay.api.base.params.ResponseParam;
import com.xpay.api.base.utils.SignUtil;
import com.xpay.common.statics.exceptions.BizException;
import com.xpay.common.util.utils.JsonUtil;
import com.xpay.common.util.utils.OkHttpUtil;
import com.xpay.common.util.utils.RSAUtil;
import com.xpay.common.util.utils.StringUtil;

/**
 * 请求处理工具类
 *
 * @author chenyf
 * @date 2018-12-15
 */
public class RequestUtil {

    /**
     * 发起请求，对于返回的Response内容
     * 说明：
     * 1、如果request中的aes_key不为空，且secretKey中的aes_key_decrypt_key不为空，会自动使用aes_key_decrypt_key为aes_key进行加密
     * 1、验签时不用理会http状态码，而只管对返回的内容进行验签，如果没有内容返回，也当做验签失败处理
     * 2、如果验签不通过，一切结果都不可信，不用理会http状态码是什么，也不用例会返回的内容是什么
     *
     * @param url
     * @param request
     * @param secretKey
     * @return
     * @throws BizException
     */
    public static ResponseParam doRequest(String url, RequestParam request, SecretKey secretKey) throws BizException {
        requestParamValid(request, secretKey);

        try {
            String signStr = SignUtil.sign(request, request.getSign_type(), secretKey.getReqSignPriKey());
            request.setSign(signStr);
        } catch (Exception e) {
            throw new BizException("签名失败: " + e.getMessage(), e);
        }

        if (StringUtil.isNotEmpty(request.getSec_key()) && StringUtil.isNotEmpty(secretKey.getSecKeyEncryptPubKey())) {
            request.setSec_key(RSAUtil.encryptByPublicKey(request.getSec_key(), secretKey.getSecKeyEncryptPubKey()));
        }

        String respJson = null;
        try {
            respJson = OkHttpUtil.postJsonSync(url, JsonUtil.toString(request)).body().string();
        } catch (Exception e) {
            throw new BizException("发送http请求时发生异常: " + e.getMessage(), e);
        }
        if (StringUtil.isEmpty(respJson)) {
            throw new BizException("请求完成，但响应信息为空");
        }

        ResponseParam response = null;
        try {
            response = JsonUtil.toBean(respJson, ResponseParam.class);
        } catch (Exception e) {
            throw new BizException("请求完成，但响应信息转换失败: " + e.getMessage() + "，respJson=" + respJson, e);
        }

        try {
            boolean isOk = SignUtil.verify(response, response.getSign(), response.getSign_type(), secretKey.getRespVerifyPubKey());
            if (!isOk) {
                throw new BizException("响应信息验签不通过！");
            }
        } catch (Exception e) {
            throw new BizException("请求完成，但响应信息验签异常: " + e.getMessage() + "，respJson=" + respJson, e);
        }

        if (StringUtil.isNotEmpty(response.getSec_key()) && StringUtil.isNotEmpty(secretKey.getSecKeyDecryptPriKey())) {
            String secKey = RSAUtil.decryptByPrivateKey(response.getSec_key(), secretKey.getSecKeyDecryptPriKey());
            response.setSec_key(secKey);
        }

        return response;
    }

    private static <T> void requestParamValid(RequestParam request, SecretKey secretKey) {
        if (request == null) {
            throw new BizException("request不能为空");
        } else if (StringUtil.isEmpty(request.getMethod())) {
            throw new BizException("Request.method不能为空");
        } else if (StringUtil.isEmpty(request.getVersion())) {
            throw new BizException("Request.version不能为空");
        } else if (request.getData() == null || StringUtil.isEmpty(request.getData().toString())) {
            throw new BizException("Request.data不能为空");
        } else if (StringUtil.isEmpty(request.getSign_type())) {
            throw new BizException("Request.sign_type不能为空");
        } else if (StringUtil.isEmpty(request.getMch_no())) {
            throw new BizException("Request.mch_no不能为空");
        }

        if (secretKey == null) {
            throw new BizException("secretKey不能为空");
        } else if (StringUtil.isEmpty(secretKey.getReqSignPriKey())) {
            throw new BizException("SecretKey.reqSignPriKey不能为空");
        } else if (StringUtil.isEmpty(secretKey.getRespVerifyPubKey())) {
            throw new BizException("SecretKey.respVerifyPubKey不能为空");
        }
    }

}

File: sdk/sdk-api-php5/utils/RequestUtil.php
<?php

namespace xpay;

use Exception;

/**
 * 发起交易请求的工具类
 * Class RequestUtil
 * @package utils
 */
class RequestUtil{

    /**
     * 发起交易请求
     *   说明：
     *      1、此方法不会对商户私钥做任何处理，理论上不会造成私钥泄漏，如果商户觉得把私钥传递到此方法不安全，可以在调用本方法前
     *          自行调用SignUtil的相关方法为请求参数生成签名，自行调用RSAUtil的相关方法对sec_key进行加解密。
     *      2、发起请求时，如果SecretKey.reqSignKey不为空，当前方法会自动为请求参数生成签名、得到系统响应之后也会自动对响应
     *          数据进行验签，签名或验签失败会抛出 SDKException 异常
     *      3、发送请求时，如果Request中的sec_key和SecretKey.secKeyEncryptKey都不为空，会自动给sec_key加密
     *      4、得到响应时，如果Response中的sec_key和SecretKey.secKeyDecryptKey都不为空，会自动给sec_key解密
     *
     * @param string $url
     * @param Request $request
     * @param SecretKey $secretKey
     * @return Response
     * @throws SDKException
     * @throws \ReflectionException
     */
    public static function doRequest($url, Request $request, SecretKey $secretKey)
    {
        //1.参数校验
        self::requestParamValid($request, $secretKey);

        if(! is_string($request->getData())){
            //注意：如果data是一个没有实现JsonSerializable接口的对象，那这个类中被private修饰的属性将不能被json_encode()获取到
            $request->setData(json_encode($request->getData(), JSON_UNESCAPED_UNICODE));
        }

        echo "原请求参数："; print_r($request);
        //2.添加签名
        if($secretKey->getReqSignKey()){
            $signStr = SignUtil::getSortedString($request, false);
            echo "请求待签名串：" . $signStr . "\n";
            $signStr = SignUtil::sign($signStr, $request->getSignType(), $secretKey->getReqSignKey());
            $request->setSign($signStr);
        }else if(! $request->getSign()){
            throw new SDKException(SDKException::BIZ_ERROR, "请对请求参数添加签名！");
        }

        //3.如果存在sec_key，则对其进行rsa加密
        if($request->getSecKey() && $secretKey->getSecKeyEncryptKey()){
            $request->setSecKey(RSAUtil::encrypt($request->getSecKey(), $secretKey->getSecKeyEncryptKey()));
        }

        //4.发起http请求
        echo "http请求参数：" . json_encode($request, JSON_UNESCAPED_UNICODE) . "\n";
        $respJson = HttpUtil::postJsonSync($url, json_encode($request, JSON_UNESCAPED_UNICODE));
        if(! $respJson){
            throw new SDKException(SDKException::BIZ_ERROR, "请求完成，但响应信息为空");
        }
        echo "http响应参数：" . $respJson . "\n";
        //5.响应结果转换
        $response = new Response();
        try{
            ObjectUtil::fillProperties($response, $respJson);
        }catch (Exception $e){
            throw new SDKException(SDKException::BIZ_ERROR, "请求完成，但响应信息转换失败: " . $e->getMessage() . "，respJson = " . $respJson);
        }

        $isOk = false;
        try{
            $signStr = SignUtil::getSortedString($response, false);
            echo "响应待签名串：" . $signStr . "\n";
            $isOk = SignUtil::verify($signStr, $response->getSign(), $response->getSignType(), $secretKey->getRespVerifyKey());
        }catch (Exception $e){
            throw new SDKException(SDKException::BIZ_ERROR, "请求完成，但响应信息验签异常: " . $e->getMessage() . "，respJson = " . $respJson);
        }
        if($isOk !== true){
            throw new SDKException(SDKException::BIZ_ERROR, "响应信息验签不通过！ respJson = " . $respJson);
        }

        if($response->getSecKey() && $secretKey->getSecKeyDecryptKey()){
            $secKey = RSAUtil::decrypt($response->getSecKey(), $secretKey->getSecKeyDecryptKey());
            $response->setSecKey($secKey);
        }
        echo "转换后的响应数："; print_r($response);
        return $response;
    }

    /**
     * 简单参数校验
     * @param Request $request
     * @param SecretKey $secretKey
     * @throws Exception
     */
    private static function requestParamValid(Request $request, SecretKey $secretKey){
        if(! $request){
            throw new SDKException(SDKException::PARAM_ERROR, "request不能为空");
        }else if(! $request->getMethod()){
            throw new SDKException(SDKException::PARAM_ERROR, "Request.method不能为空");
        }else if(! $request->getVersion()){
            throw new SDKException(SDKException::PARAM_ERROR, "Request.version不能为空");
        }else if(! $request->getData()){
            throw new SDKException(SDKException::PARAM_ERROR, "Request.data不能为空");
        }else if(! $request->getSignType()){
            throw new SDKException(SDKException::PARAM_ERROR, "Request.sign_type不能为空");
        }else if(! $request->getMchNo()){
            throw new SDKException(SDKException::PARAM_ERROR, "Request.mch_no不能为空");
        }

        if(! $secretKey){
            throw new SDKException(SDKException::PARAM_ERROR, "secretKey不能为空");
        }else if(! $secretKey->getRespVerifyKey()){
            throw new SDKException(SDKException::PARAM_ERROR, "SecretKey.respVerifyKey不能为空");
        }
    }
}

